specs/_features/eip6110/beacon-chain.md

```diff
@@ -0,0 +1,429 @@
+    for_ops(body.attestations, process_attestation)
+    for_ops(body.deposits, process_deposit)  # [Modified in EIP-6110]
+    for_ops(body.voluntary_exits, process_voluntary_exit)
+    for_ops(body.bls_to_execution_changes, process_bls_to_execution_change)
+
+    # [New in EIP-6110]
+    if is_execution_enabled(state, body):
+        for_ops(body.execution_payload.deposit_receipts, process_deposit_receipt)
+```
+
+#### New `get_validator_from_deposit_receipt`
+
+```python
+def get_validator_from_deposit_receipt(deposit_receipt: DepositReceipt) -> Validator:
+    amount = deposit_receipt.amount
+    effective_balance = min(amount - amount % EFFECTIVE_BALANCE_INCREMENT, MAX_EFFECTIVE_BALANCE)
+
+    return Validator(
+        pubkey=deposit_receipt.pubkey,
+        withdrawal_credentials=deposit_receipt.withdrawal_credentials,
+        activation_eligibility_epoch=FAR_FUTURE_EPOCH,
+        activation_epoch=FAR_FUTURE_EPOCH,
+        exit_epoch=FAR_FUTURE_EPOCH,
+        withdrawable_epoch=FAR_FUTURE_EPOCH,
+        effective_balance=effective_balance,
+    )
+```
+
+#### New `process_deposit_receipt`
+
+```python
+def process_deposit_receipt(state: BeaconState, deposit_receipt: DepositReceipt) -> None:
+    # Set deposit receipt start index
+    if state.deposit_receipts_start_index == NOT_SET_DEPOSIT_RECEIPTS_START_INDEX:
+        state.deposit_receipts_start_index = deposit_receipt.index
+
+    # Signify the end of transition to in-protocol deposit logic
+    if state.eth1_deposit_index >= state.deposit_receipts_start_index:
+        state.eth1_deposit_index = deposit_receipt.index + 1
+
+    pubkey = deposit_receipt.pubkey
+    amount = deposit_receipt.amount
+    validator_pubkeys = [validator.pubkey for validator in state.validators]
+    if pubkey not in validator_pubkeys:
+        # Verify the deposit signature (proof of possession) which is not checked by the deposit contract
+        deposit_message = DepositMessage(
+            pubkey=deposit_receipt.pubkey,
+            withdrawal_credentials=deposit_receipt.withdrawal_credentials,
+            amount=deposit_receipt.amount,
+        )
+        domain = compute_domain(DOMAIN_DEPOSIT)  # Fork-agnostic domain since deposits are valid across forks
+        signing_root = compute_signing_root(deposit_message, domain)
+        # Initialize validator if the deposit signature is valid
+        if bls.Verify(pubkey, signing_root, deposit_receipt.signature):
+            state.validators.append(get_validator_from_deposit_receipt(deposit_receipt))
+            state.balances.append(amount)
+            state.previous_epoch_participation.append(ParticipationFlags(0b0000_0000))
+            state.current_epoch_participation.append(ParticipationFlags(0b0000_0000))
+            state.inactivity_scores.append(uint64(0))
+    else:
+        # Increase balance by deposit amount
+        index = ValidatorIndex(validator_pubkeys.index(pubkey))
+        increase_balance(state, index, amount)
```

Comment:
Is this code snippet identical to `process_deposit` content except for using `deposit_receipt: DepositReceipt` rather than `deposit: Deposit`?

It may be worth some refactoring to avoid repeated code:

```python
def handle_deposit(state: BeaconState, deposit: PyUnion[Deposit, DepositDepositReceipt]) -> None:
    pubkey = deposit.pubkey
    ...
```

Need to check:
- `handle_*` is a new helper naming here. Any better ideas?
- We didn't use `PyUnion` in the specs. It's the Python Typing `from typing import Union,` and I renamed it to `PyUnion` to distinguish it from SSZ Union.

Reply:
I was thinking about introducing the following method:
```python
def apply_deposit(state: BeaconState, pubkey: BLSPubkey, withdrawal_credentials: Bytes32, amount: uint64) -> None
```

But hesitated to change already finalized pieces of the spec.

Reply:
Or you mean to reuse the same function in this particular spec doc? Sounds reasonable!

Reply:
Oh yes, it looks better than I thought! 👍